espnet2/train/trainer.py

```diff
@@ -0,0 +1,292 @@
+        all_steps_are_invalid = True
+        for iiter, (_, batch) in enumerate(iterator, 1):
+            assert isinstance(batch, dict), type(batch)
+            batch = to_device(batch, "cuda" if ngpu > 0 else "cpu")
+            if no_forward_run:
+                all_steps_are_invalid = False
+                reporter.register({})
+                continue
+
+            if ngpu <= 1:
+                # NOTE(kamo): data_parallel also should work with ngpu=1,
+                # but for debuggability it's better to keep this block.
+                loss, stats, weight = model(**batch)
+            else:
+                loss, stats, weight = data_parallel(
+                    model, (), range(ngpu), module_kwargs=batch
+                )
+                # Weighted averaging of loss from torch-data-parallel
+                loss = (loss * weight.to(loss.dtype)).sum(0) / weight.sum()
+                stats = {
+                    k: (v * weight.to(v.dtype)).sum(0) / weight.sum()
+                    if v is not None
+                    else None
+                    for k, v in stats.items()
+                }
+                weight = weight.sum()
+            reporter.register(stats, weight)
+
+            if train_dtype in ("O0", "O1", "O2", "O3"):
+                from apex import amp
+
+                with amp.scale_loss(loss, optimizers) as scaled_loss:
+                    scaled_loss.backward()
+            else:
+                loss.backward()
+
+            # gradient noise injection
+            if grad_noise:
+                add_gradient_noise(
+                    model,
+                    reporter.get_total_count(),
+                    duration=100,
+                    eta=1.0,
+                    scale_factor=0.55,
+                )
+
+            # compute the gradient norm to check if it is normal or not
+            grad_norm = torch.nn.utils.clip_grad_norm_(model.parameters(), grad_clip)
+            if iiter % accum_grad == 0:
+                if not np.isfinite(grad_norm):
+                    logging.warning(
+                        f"The grad norm is {grad_norm}. Skipping updating the model."
+                    )
+                else:
+                    all_steps_are_invalid = False
+                    optimizer.step()
+                optimizer.zero_grad()
+                if isinstance(scheduler, AbsBatchStepScheduler):
+                    scheduler.step()
+
+                # Register lr
+                reporter.register(
+                    {
+                        f"lr_{i}": pg["lr"]
+                        for i, pg in enumerate(optimizer.param_groups)
+                        if "lr" in pg
+                    },
+                    not_increment_count=True,
+                )
+
+            if iiter % log_interval == 0:
+                reporter.logging(nlatest=log_interval)
+        return all_steps_are_invalid
+
+    @classmethod
+    @torch.no_grad()
+    def eval_one_epoch(
```

Comment:
if we want to perform the evaluation for every `n` iterations, do we need to modify this part?
this trainer seems to assume epoch-based processing too much, and we'd have a lot of modifications in the future.
the chainer-based trainer has a lot of flexibility for such perspective, but this makes the trainer very complicated...

Reply:
Your pointing out may be important. Could you tell me another idea of the non-epoch-based processes if you already have?

As related topics to it, we might need to redesign `AbsTask.run()`. I assumed that the epoch process are consists of `train()`, `eval()`, or `plot_attention()`, and another block can't be inserted, e.g. `train2()`, supervision generation, data cleaning by model, or changing model structure adoptively?? (I think this is not current matters.)

Reply:
If we augment the data (speed perturbation, process various RIRs, or adding noises), one epoch of the amount of training data would become several times of that of the original data. Then, epoch-based monitoring is too coarse. People often use iteration-based monitoring for this setup. Our method could perform such data augmentation dynamically and may still keep original epoch size though. This is just a note.

Reply:
As you know, the Chainer trainer has a trigger function to control it.

Reply:
### Idea1: Evaluate in the middle of training iterations.

```python
for epoch in range(max_epoch):
    for batch in enumerate(train_loader):
        train_one_batch(batch, reporter)
        if reporter.total_number_of_iterations() % eval_interval == 0:
            evaluate_one_epoch(eval_iter)
            plot_attention(eval_iter)
            scheduler_step()
            save_snapshot()
            check_early_stopping()
```

### Idea2: Use subset of training data in every epochs.

```python
for epoch in range(max_epoch):
   train_one_epoch(train_iter.generate_subset_iterator_randomly())
   evaluate_one_epoch(eval_iter)
   ...
```

The latter can't guaranteed using whole dataset, but seems natural as bootstrap method.

Reply:
Regarding resuming of training: Unlike chainer's iterator, Pytorch dataloader doesn't provide the system for resuming from the middle of epoch, so if we'll go to idea1's direction, we need to save the number of iterations additionally and seek the batch-sampler at the staring time . 

And, as related topics to it, currently the random-seed is not considered for resuming. I know fairseq carefully handle it for data-loader. i.e. fairseq changes the random seed according to the epoch number. I'm not sure it is worth to do.

Reply:
@sw005320 Could you decide the direction? We need a little big modification for the iterator structure with idea1. Idea2 is more simple implementation.

Reply:
Anyway it should be done in next PR.

Reply:
Recently I made the trainer which is similar to the idea 1. So +1 for idea 1.

https://github.com/kan-bayashi/ParallelWaveGAN/blob/2908bc5bce95f903f27a015f172ffd1b20017560/parallel_wavegan/bin/train.py#L208

Reply:
Sorry for my late reply.
Now, I'm back to normal.
I think idea 1 is intuitive.

Reply:
O.K. I already have gone to the direction.